features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/composer/VoiceMessageComposerPlayer.kt

```diff
@@ -17,96 +17,233 @@
+            playState = calcPlayState(prevState.playState, seekingTo, state),
             currentPosition = state.currentPosition,
-            duration = state.duration ?: 0L,
+            duration = state.duration,
+            seekingTo = seekingTo,
+        )
+    }.map {
+        State(
+            playState = it.playState,
+            currentPosition = it.currentPosition,
+            progress = calcProgress(it),
         )
     }.distinctUntilChanged()
 
+    /**
+     * Set the voice message to be played.
+     */
+    suspend fun setMedia(mediaPath: String) {
+        this.mediaPath = mediaPath
+        mediaPlayer.setMedia(
+            uri = mediaPath,
+            mediaId = mediaPath,
+            mimeType = MIME_TYPE,
+        )
+    }
+
     /**
      * Start playing from the current position.
      *
-     * @param mediaPath The path to the media to be played.
-     * @param mimeType The mime type of the media file.
+     * Call [setMedia] before calling this method.
      */
-    suspend fun play(mediaPath: String, mimeType: String) {
-        if (mediaPath == curPlayingMediaId) {
-            mediaPlayer.play()
-        } else {
-            lastPlayedMediaPath = mediaPath
-            mediaPlayer.setMedia(
-                uri = mediaPath,
-                mediaId = mediaPath,
-                mimeType = mimeType,
-            )
-            mediaPlayer.play()
+    suspend fun play() {
+        val mediaPath = this.mediaPath
+        if (mediaPath == null) {
+            Timber.e("Set media before playing")
+            return
         }
+
+        mediaPlayer.ensureMediaReady(mediaPath)
+
+        mediaPlayer.play()
     }
 
     /**
      * Pause playback.
      */
     fun pause() {
-        if (lastPlayedMediaPath == curPlayingMediaId) {
+        if (mediaPath == mediaPlayer.state.value.mediaId) {
             mediaPlayer.pause()
         }
     }
 
+    /**
+     * Seek to a given position in the current media.
+     *
+     * Call [setMedia] before calling this method.
+     *
+     * @param position The position to seek to between 0 and 1.
+     */
+    suspend fun seek(position: Float) {
+        val mediaPath = this.mediaPath
+        if (mediaPath == null) {
+            Timber.e("Set media before seeking")
+            return
+        }
+
+        seekJob?.cancelAndJoin()
+        seekingTo.value = position
+        seekJob = coroutineScope.launch {
+            val mediaState = mediaPlayer.ensureMediaReady(mediaPath)
+            val duration = mediaState.duration ?: return@launch
+            val positionMs = (duration * position).toLong()
+            mediaPlayer.seekTo(positionMs)
+        }.apply {
+            invokeOnCompletion {
+                seekingTo.value = null
+            }
+        }
+    }
+
+    private suspend fun MediaPlayer.ensureMediaReady(mediaPath: String): MediaPlayer.State {
+        val state = state.value
+        if (state.mediaId == mediaPath && state.isReady) {
```

Comment:
Sorry about not checking for this properly in the MediaPlayer. I've opened this: https://github.com/vector-im/element-x-android/pull/1783